src/algebra/char_p/basic.lean

```diff
@@ -328,17 +328,18 @@ namespace char_p
 section
 variables [ring R]
 
-lemma char_p_to_char_zero [char_p R 0] : char_zero R :=
+lemma char_p_to_char_zero (R : Type*) [add_left_cancel_monoid R] [has_one R] [char_p R 0] :
```

Comment:
Possible safest to generalize less, as
```suggestion
lemma char_p_to_char_zero (R : Type*) [add_cancel_monoid R] [has_one R] [char_p R 0] :
```

Reply:
Eric, I saw your comment only now!  I can create a new PR, changing the assumption.  However, as this is a lemma, not an instance, what is it that worries you?

Reply:
Just that we can end up in a weird position where half of mathlib is generalized to left-cancellation and half to rjght-cancellation (in cases where one or the other is sufficient), and it doesn't really benefit anyone and makes things confusing.

It's probably not worth creating a new PR over

Reply:
If `mul_opposite` works as I think it should, a "symmetric" result that holds with `left` assumptions, also holds with `right` assumptions.  It is tricky to formalize this "meta-result", but in any concrete case, it should be easy to do it.

A `to_right` attribute, analogous to `to_additive` might help...